### PR TITLE
Check module visibility in method handles

### PR DESCRIPTION
**Pull request by pdbain-ibm**
Ensure target classes, paramater types, and return types are visible.

Signed-off-by: Peter Bain <peter_bain@ca.ibm.com>

### CODE REVIEW

**Comment by pdbain-ibm** [comment]
FYI @DanHeidinga @pshipton

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 1]
```diff
@@ -47,7 +47,7 @@
 import com.ibm.oti.vm.VM;
 import com.ibm.oti.vm.VMLangAccess;
 
-/*[IF Sidecar19-SE]
+/*[IF Sidecar19-SE]*/
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: why this change?

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 2]
```diff
@@ -634,14 +636,26 @@ public MethodHandle findStatic(Class<?> clazz, String methodName, MethodType typ
 		 * @param methodName - the name of the method
 		 * @param type - the type of the method
 		 * @return a MethodHandle that will do virtual dispatch on the first argument
-		 * @throws IllegalAccessException - if method is static or access is refused
+		 * @throws IllegalAccessException - if method is static or access is refused or clazz's package is not exported to the current module
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: This javadoc will be generated for both 8 & 9+.  Module should only be mentioned for 9

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 3]
```diff
@@ -634,14 +636,26 @@ public MethodHandle findStatic(Class<?> clazz, String methodName, MethodType typ
 		 * @param methodName - the name of the method
 		 * @param type - the type of the method
 		 * @return a MethodHandle that will do virtual dispatch on the first argument
-		 * @throws IllegalAccessException - if method is static or access is refused
+		 * @throws IllegalAccessException - if method is static or access is refused or clazz's package is not exported to the current module 
+		 * @throws IllegalAccessError - if the type's packages are not exported to the current module
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: module mentioned in 8

**Comment by DanHeidinga** [comment]
The PR only adds this check to `findVirtual`.  What about the other `find{static,special,bind,setter,getter}` methods?  Do they need the module access checks as well?

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 4]
```diff
@@ -634,14 +636,26 @@ public MethodHandle findStatic(Class<?> clazz, String methodName, MethodType typ
...
+		 * @throws IllegalAccessError - if the type's packages are not exported to the current module 
 		 * @throws NullPointerException - if clazz, methodName or type is null
 		 * @throws NoSuchMethodException - if clazz has no virtual method named methodName with signature matching type
 		 * @throws SecurityException - if any installed SecurityManager denies access to the method
 		 */
 		public MethodHandle findVirtual(Class<?> clazz, String methodName, MethodType type) throws IllegalAccessException, NoSuchMethodException {
 			nullCheck(clazz, methodName, type);
 			
+			/*[IF Sidecar19-SE]*/
+			if (INTERNAL_PRIVILEGED != accessMode) {
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: shouldn't this be added to `checkAccess()`?

**Comment by pdbain-ibm** [comment]
> What about the other find{static,special,bind,setter,getter} methods?

Possibly. We don't have tests for these yet.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 2 above)
Comment (pdbain-ibm) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: Will fix.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 1 above)
Comment (DanHeidinga) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: Doesn't that just move the syntax error from 9 to 8?

Comment (pdbain-ibm) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: No: this change eliminates most of the syntax errors in the unprocessed code.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 5]
```diff
@@ -392,6 +394,17 @@ void checkAccess(Class<?> clazz) throws IllegalAccessException {
 		private void checkAccess(Class<?> targetClass, String name, int memberModifiers, MethodHandle handle) throws IllegalAccessException {
 			checkClassAccess(targetClass);
 
+			/*[IF Sidecar19-SE]*/
+			if (null != handle) {
+				MethodType type = handle.type();
+				Module accessModule = accessClass.getModule();
+				checkClassModuleVisibility(accessModule, targetClass, false);
+				checkClassModuleVisibility(accessModule, type.returnType, true);
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: There are two calls to `checkClassModuleVisibility`, one setting `throwError` to false while the other says true.  Which is intended?

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 6]
```diff
@@ -681,6 +717,32 @@ public MethodHandle findVirtual(Class<?> clazz, String methodName, MethodType ty
...
+		static void checkClassModuleVisibility(Module accessModule, Class<?> targetClass, boolean throwError) 
+				throws IllegalAccessError, IllegalAccessException {
+			Module targetModule = targetClass.getModule();
+			if (!Objects.equals(accessModule, targetModule)
+					&& !(targetModule.isExported(targetClass.getPackageName(), accessModule)
+							&& accessModule.canRead(targetModule))) {
+				/*[MSG "K0664", "{0} is not in an exported package"]*/
+				String message = Msg.getString("K0664", targetClass.getPackageName()); //$NON-NLS-1$
+				if (throwError) {
+					throw new IllegalAccessError(message);
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: likely better to only throw one exception and let the code that needs the Error wrap the Exception.

File: jcl/src/java.base/share/classes/com/ibm/oti/util/ExternalMessages-MasterIndex.properties [hunk 7]
```diff
@@ -1160,6 +1160,7 @@ K0660="Internal error while obtaining private constructor."
 K0661="Internal error while obtaining GcInfo instance."
 K0662="maxDepth must not be negative."
 K0663="Invalid or unsupported dump agent option, cannot be triggered."
+K0664="{0} is not in an exported package"
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/com/ibm/oti/util/ExternalMessages-MasterIndex.properties]: Is there no existing message that can be reused?

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 8]
```diff
@@ -392,6 +394,17 @@ void checkAccess(Class<?> clazz) throws IllegalAccessException {
 		private void checkAccess(Class<?> targetClass, String name, int memberModifiers, MethodHandle handle) throws IllegalAccessException {
 			checkClassAccess(targetClass);
 
+			/*[IF Sidecar19-SE]*/
+			if (null != handle) {
+				MethodType type = handle.type();
+				Module accessModule = accessClass.getModule();
+				checkClassModuleVisibility(accessModule, targetClass, false);
+				checkClassModuleVisibility(accessModule, type.returnType, true);
+				for (Class<?> c: type.parameterArray()) {
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: `type.parameterArray()` allocates a new array - since this code is in the j.l.i package and we know it won't modify the array, better to use `type.arguments` directly

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 9]
```diff
@@ -681,6 +717,32 @@ public MethodHandle findVirtual(Class<?> clazz, String methodName, MethodType ty
...
+		/*[IF Sidecar19-SE]*/
+		/**
+		 * Check if targetClass is in a package visible from the accessModule
+		 * @param accessModule module of the referring class
+		 * @param targetClass Class which the referring class is accessing
+		 * @param throwError Select the throwable thrown if the targetClass is not visible
+		 * @throws IllegalAccessError thrown if throwError is true
+		 * @throws IllegalAccessException thrown if throwError is false
+		 */
+		static void checkClassModuleVisibility(Module accessModule, Class<?> targetClass, boolean throwError)
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: Please put the throw clauses on the same line.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 10]
```diff
@@ -681,6 +717,32 @@ public MethodHandle findVirtual(Class<?> clazz, String methodName, MethodType ty
...
+		 * @throws IllegalAccessError thrown if throwError is true
+		 * @throws IllegalAccessException thrown if throwError is false
+		 */
+		static void checkClassModuleVisibility(Module accessModule, Class<?> targetClass, boolean throwError) 
+				throws IllegalAccessError, IllegalAccessException {
+			Module targetModule = targetClass.getModule();
+			if (!Objects.equals(accessModule, targetModule)
+					&& !(targetModule.isExported(targetClass.getPackageName(), accessModule)
+							&& accessModule.canRead(targetModule))) {
+				/*[MSG "K0664", "{0} is not in an exported package"]*/
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: Can you add the two modules to this message?  It will be easier for a user to resolve an access failure if they get the module info in the exception.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 11]
```diff
@@ -681,6 +717,32 @@ public MethodHandle findVirtual(Class<?> clazz, String methodName, MethodType ty
...
+		 * @param accessModule module of the referring class
+		 * @param targetClass Class which the referring class is accessing
+		 * @param throwError Select the throwable thrown if the targetClass is not visible
+		 * @throws IllegalAccessError thrown if throwError is true
+		 * @throws IllegalAccessException thrown if throwError is false
+		 */
+		static void checkClassModuleVisibility(Module accessModule, Class<?> targetClass, boolean throwError) 
+				throws IllegalAccessError, IllegalAccessException {
+			Module targetModule = targetClass.getModule();
+			if (!Objects.equals(accessModule, targetModule)
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: Can you rewrite this as positive conditions?

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 5 above)
Comment (pdbain-ibm) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: The throwable depends on the context.  If the target class is not visible, we need to throw an IllegalAccessException. If the parameter or return types are not visible, we need to throw a IllegalAccessError.

File: jcl/src/java.base/share/classes/com/ibm/oti/util/ExternalMessages-MasterIndex.properties (hunk 7 above)
Comment (pdbain-ibm) [reply jcl/src/java.base/share/classes/com/ibm/oti/util/ExternalMessages-MasterIndex.properties]: Yes.  Fixed.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 9 above)
Comment (pdbain-ibm) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: Okay.

**Comment by pdbain-ibm** [comment]
Passes designer tests.  Will run a larger regression suite tomorrow.

**Comment by pdbain-ibm** [comment]
JCKL tests pass.  @DanHeidinga or @pshipton could you take another look at the code?

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 12]
```diff
@@ -625,6 +658,23 @@ public MethodHandle findStatic(Class<?> clazz, String methodName, MethodType typ
...
+		 * Return a MethodHandle to a virtual method.  The method will be looked up in the first argument 
+		 * (aka receiver) prior to dispatch.  The type of the MethodHandle will be that of the method
+		 * with the receiver type prepended.
+		 * 
+		 * @param clazz - the class defining the method
+		 * @param methodName - the name of the method
+		 * @param type - the type of the method
+		 * @return a MethodHandle that will do virtual dispatch on the first argument
+		 * @throws IllegalAccessException - if method is static or access is refused or clazz's package is not exported to the current module 
+		 * @throws IllegalAccessError - if the type's packages are not exported to the current module
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: Has the RI thrown IA**Error** from findVirtual?

The convention is usually that the VM throws IllegalAccessError and reflective style operations throw Exceptions.  We're using the same code to do double duty as the VM is calling into Java for these lookups.

I would expect exception / error conversion to happen at the boundaries.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 13]
```diff
@@ -512,10 +533,22 @@ private void checkClassAccess(Class<?> targetClass) throws IllegalAccessExceptio
...
 				}
 			}
-			
+
 			/*[MSG "K0587", "'{0}' no access to: '{1}'"]*/
 			throw new IllegalAccessException(com.ibm.oti.util.Msg.getString("K0587", accessClass.getName(), targetClass.getName()));  //$NON-NLS-1$
 		}
+
+		void checkLookupVisibility(Class<?> targetClass) throws IllegalAccessException, IllegalAccessError {
+			if (this.equals(PUBLIC_LOOKUP)) {
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: Rather than checking is identically PUBLIC_LOOKUP, we should check for the `UNCONDITIONAL` mode access bit - see https://docs.oracle.com/javase/9/docs/api/java/lang/invoke/MethodHandles.html#publicLookup--

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 14]
```diff
@@ -681,6 +732,25 @@ public MethodHandle findVirtual(Class<?> clazz, String methodName, MethodType ty
...
 		}
+
+		/*[IF Sidecar19-SE]*/
+		/**
+		 * Check if targetClass is in a package visible from the accessModule
+		 * @param accessModule module of the referring class
+		 * @param targetClass Class which the referring class is accessing
+		 * @throws IllegalAccessException if the targetClass is not visible
+		 */
+		static void checkClassModuleVisibility(Module accessModule, Class<?> targetClass) throws IllegalAccessError, IllegalAccessException {
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: This probably also needs to take the `UNCONDITIONAL` access bit into account (ie: public lookup)

**Comment by pdbain-ibm** [comment]
The current change causes a regression in another test.  I am investigating.

**Comment by pdbain-ibm** [comment]
Regression was due to stale base code.  
@DanHeidinga or @pshipton would you kindly review and run a PR build?

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 12 above)
Comment (pdbain-ibm) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: > Has the RI thrown IAError from findVirtual?

Yes.  This matches the behaviour of the reference implementation.

**Comment by pdbain-ibm** [comment]
@DanHeidinga or @pshipton do you have any comments on this PR?  If not, could you
please run a sanity build?

Thank you.

**Comment by DanHeidinga** [comment]
Jenkins test sanity plinux jdk8,jdk9
